internals/overlord/logstate/opentelemetry/opentelemetry.go

```diff
@@ -0,0 +1,405 @@
+	TimeUnixNano uint64 `json:"timeUnixNano"`
+	// The severity text (also known as log level). The original string representation as
+	// it is known at the source. [Optional].
+	SeverityText string `json:"severityText"`
+	// Numerical value of the severity, normalized to values described in Log Data Model.
+	// [Optional].
+	SeverityNumber int `json:"severityNumber"`
+	// A value containing the body of the log record. Can be for example a human-readable
+	// string message (including multi-line) describing the event in a free form or it can
+	// be a structured data composed of arrays and maps of other values. [Optional].
+	Body AttributeValue `json:"body"`
+	// Additional attributes that describe the specific event occurrence. [Optional].
+	// Attribute keys MUST be unique (it is not allowed to have more than one
+	// attribute with the same key).
+	Attributes []KeyValue `json:"attributes,omitempty"`
+}
+
+type Client struct {
+	options    *ClientOptions
+	target     *plan.LogTarget
+	httpClient *http.Client
+
+	// To store log entries, keep a buffer of size 2*MaxRequestEntries with a
+	// sliding window "entries" of size MaxRequestEntries.
+	buffer  []otelEntryWithService
+	entries []otelEntryWithService
+
+	// Store the custom labels for each service (resource attributes in OTEL).
+	resourceAttributes map[string][]KeyValue
+}
+
+func NewClient(target *plan.LogTarget) *Client {
+	return NewClientWithOptions(target, &ClientOptions{})
+}
+
+// ClientOptions allows overriding default parameters (e.g. for testing).
+type ClientOptions struct {
+	RequestTimeout    time.Duration
+	MaxRequestEntries int
+	BatchSize         int
+}
+
+func NewClientWithOptions(target *plan.LogTarget, options *ClientOptions) *Client {
+	options = fillDefaultOptions(options)
+	c := &Client{
+		options:            options,
+		target:             target,
+		httpClient:         &http.Client{Timeout: options.RequestTimeout},
+		buffer:             make([]otelEntryWithService, 2*options.MaxRequestEntries),
+		resourceAttributes: make(map[string][]KeyValue),
+	}
+	// c.entries should be backed by the same array as c.buffer.
+	c.entries = c.buffer[:0]
+	return c
+}
+
+func fillDefaultOptions(options *ClientOptions) *ClientOptions {
+	if options.RequestTimeout == 0 {
+		options.RequestTimeout = requestTimeout
+	}
+	if options.MaxRequestEntries == 0 {
+		options.MaxRequestEntries = maxRequestEntries
+	}
+	if options.BatchSize == 0 {
+		options.BatchSize = batchSize
+	}
+	return options
+}
+
+// SetLabels sets resource attributes for a service. Labels are analogous to OpenTelemetry's resource attributes.
+func (c *Client) SetLabels(serviceName string, attributes map[string]string) {
+	if attributes == nil {
+		delete(c.resourceAttributes, serviceName)
+		return
+	}
+
+	// Convert attributes to KeyValue format.
+	keyValuePairs := make([]KeyValue, 0, len(attributes)+1)
+
+	// Add service.name attribute.
+	keyValuePairs = append(keyValuePairs, KeyValue{
+		Key:   "service.name",
```

Comment:
This is called `pebble_service` in the Loki labels. Would it be good to be consistent with that, or is `service.name` a standard OpenTelemetry thing?